File: Python3/0059_spiral_matrix_ii.py
# https://leetcode.cn/problems/spiral-matrix-ii/

class Solution:
    def generateMatrix(self, n):
        """
        Inputs:
            n: int
        Outputs:
            res: list[list[int]]
        """
        res = [[0 for _ in range(n)] for _ in range(n)]
        dirs = [[0,1], [1,0], [0,-1], [-1,0]]
        dir_id = 0
        dir = dirs[dir_id]
        row, top, bottom = 0, 0, n-1
        col, left, right = 0, 0, n-1
        for val in range(1, n*n+1):
            res[row][col] = val

            i, j = row + dir[0], col + dir[1]
            if (i >= top and i <= bottom and j >= left and j <= right):
                row, col = i, j
                continue

            dir_id = (dir_id + 1) % 4
            dir = dirs[dir_id]
            row += dir[0]
            col += dir[1]
            if i < top:
                left += 1
            elif i > bottom:
                right -= 1
            elif j < left:
                bottom -= 1
            else:
                top += 1
        return res

File: Cpp/0059_spiral_matrix_ii.cpp
// https://leetcode.cn/problems/spiral-matrix-ii/

#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    vector<vector<int>> generateMatrix(int n) {
        using PII = pair<int, int>;

        vector<vector<int>> res(n, vector<int>(n));

        vector<PII> dirs = {make_pair(0,1), make_pair(1,0), make_pair(0,-1), make_pair(-1,0)};
        int dir_id = 0;
        auto dir = dirs[dir_id];
        int row = 0, top = 0, bottom = n-1;
        int col = 0, left = 0, right = n-1;
        for (int val = 1; val <= n*n; ++val) {
            res[row][col] = val;

            int i = row + dir.first, j = col + dir.second;
            if (i >= top && i <= bottom && j >= left && j <= right) {
                row = i;
                col = j;
                continue;
            }

            dir_id = (dir_id + 1) % 4;
            dir = dirs[dir_id];
            row += dir.first;
            col += dir.second;
            if (i < top) left += 1;
            else if (i > bottom) right -= 1;
            else if (j > right) top += 1;
            else bottom -= 1;
        }
        return res;
    }
};

int main()
{
    Solution solu;
    int n = 4;

    vector<vector<int>> res = solu.generateMatrix(n);
    for (vector<int> row: res) {
        for (int num: row) {
            cout << num << ", ";
        }
        cout << '\n';
    }
    return 0;
}
